tools/c7n_azure/tests/test_keyvault.py

```diff
@@ -72,3 +75,82 @@ def test_compare_permissions(self):
         p1 = {"keys": ['get'], "secrets": ['get'], "certificates": ['get']}
         p2 = {}
         self.assertFalse(WhiteListFilter.compare_permissions(p1, p2))
+
+    @arm_template('keyvault.json')
+    @patch('c7n_azure.session.Session.get_tenant_id', return_value=DEFAULT_SUBSCRIPTION_ID)
+    def test_whitelist(self, get_tenant_id):
+        """Tests basic whitelist functionality"""
+        p = self.load_policy({
+            'name': 'test-key-vault',
+            'resource': 'azure.keyvault',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cckeyvault1*'},
+                {'not': [
+                    {'type': 'whitelist',
+                     'key': 'principalName',
+                     'users': ['account1@sample.com']}
+                ]}
+            ]
+        })
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
+
+    @arm_template('keyvault-no-policies.json')
+    def test_whitelist_zero_access_policies(self):
+        """Tests that a keyvault with 0 access policies is processed properly
+        and doesn't raise an exception.
+        """
+        p = self.load_policy({
+            'name': 'test-key-vault',
+            'resource': 'azure.keyvault',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cckeyvault2*'},
+                {'not': [
+                    {'type': 'whitelist',
+                     'key': 'principalName',
+                     'users': ['account1@sample.com']}
+                ]}
+            ]
+        })
+        resources = p.run()
+        self.assertEqual(len(resources), 0)
+
+    @arm_template('keyvault.json')
+    @patch('c7n_azure.session.Session.get_tenant_id', return_value=DEFAULT_SUBSCRIPTION_ID)
+    def test_whitelist_not_authorized(self, get_tenant_id):
+        """Tests that more detailed error messaging is returned for missing and/or incorrect
+        keys regarding whitelist filtering of keyvaults based on access policies.
+        """
+        p = self.load_policy({
+            'name': 'test-key-vault',
+            'resource': 'azure.keyvault',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cckeyvault1*'},
+                {'not': [
+                    {'type': 'whitelist',
+                     'key': 'principalName',
+                     'users': ['account1@sample.com']}
+                ]}
+            ]
+        })
+
+        with self.assertRaises(KeyError) as e:
```

Comment:
I don't see how this policy is different from policy in test_whitelist(). Are there any extra steps to record the cassette?

Reply:
Added additional comments on how to regenerate cassette.